blocks/autocompleters/index.js

```diff
@@ -51,55 +63,52 @@ import BlockIcon from '../block-icon';
  * @typedef {Object} Completer
  * @property {?String} className A class to apply to the popup menu.
  * @property {String} triggerPrefix the prefix that will display the menu.
- * @property {FnGetOptions} getOptions get the block options in a resolved promise.
+ * @property {FnGetOptions} getOptions get the completer options in a resolved promise.
+ * @property {?FnGetOptionKeywords} getOptionKeywords get the keywords for a given option.
+ * @property {FnRenderOptionLabel} renderOptionLabel render the label for a given option.
  * @property {?FnAllowNode} allowNode filter the allowed text nodes in the autocomplete.
  * @property {?FnAllowContext} allowContext filter the context under which the autocomplete activates.
- * @property {FnOnSelect} onSelect
+ * @property {FnGetOptionResult} getOptionResult
  */
 
 /**
  * Returns an "completer" definition for selecting from available blocks to replace the current one.
  * The definition can be understood by the Autocomplete component.
  *
- * @param {Function} onReplace  Callback to replace the current block.
- *
  * @return {Completer} Completer object used by the Autocomplete component.
  */
-export function blockAutocompleter( { onReplace } ) {
-	// Prioritize common category in block type options
-	const options = sortBy(
-		getBlockTypes(),
-		( { category } ) => 'common' !== category
-	).map( ( blockType ) => {
-		const { name, title, icon, keywords = [] } = blockType;
-		return {
-			value: name,
-			label: [
-				<BlockIcon key="icon" icon={ icon } />,
-				title,
-			],
-			keywords: [ ...keywords, title ],
-		};
-	} );
-
-	const getOptions = () => Promise.resolve( options );
+export function blockAutocompleter() {
+	const getOptions = once( () => Promise.resolve(
+		// Prioritize common category in block type options
+		sortBy(
+			getBlockTypes(),
+			( { category } ) => 'common' !== category
+		)
+	) );
+	const getOptionKeywords = ( { title, keywords = [] } ) => [ ...keywords, title ];
+	const renderOptionLabel = ( { icon, title } ) => [
+		<BlockIcon key="icon" icon={ icon } />,
+		title,
+	];
 
 	const allowContext = ( before, after ) => {
 		return ! ( /\S/.test( before.toString() ) || /\S/.test( after.toString() ) );
 	};
 
-	const onSelect = ( blockName ) => {
-		onReplace( createBlock( blockName ) );
-	};
+	const getOptionResult = blockData => createBlock( blockData.name );
 
 	return {
+		type: 'block',
```

Comment:
That's a good question. My intent was to mark completers with what they complete, but it might be a good idea to provide an ID instead. I could see cases where someone wants to override a core autocompleter but not one added by a plugin.

What do you think?

Reply:
Yeah, that's a good point. Something that identifies this specific autocompleter, e.g. `name: 'core-autocompleter/block'` works for me.